multisense_ros/src/multisense.cpp

```diff
@@ -0,0 +1,1651 @@
+    output.dropped_messages = status.client_network.dropped_messages;
+    output.invalid_packets = status.client_network.invalid_packets;
+
+    status_pub_->publish(std::make_unique<multisense_msgs::msg::Status>(std::move(output)));
+}
+
+size_t MultiSense::num_subscribers(const rclcpp::Node::SharedPtr node, const std::string &topic)
+{
+    return num_subscribers(node.get(), topic);
+}
+
+size_t MultiSense::num_subscribers(const rclcpp::Node* node, const std::string &topic)
+{
+    const std::string full_topic = node->get_sub_namespace().empty() ? topic : node->get_sub_namespace()  + "/" + topic;
+
+    if (!rclcpp::ok())
+    {
+        return 0;
+    }
+
+    return node->count_subscribers(full_topic) ;
+}
+
+void MultiSense::initialize_parameters(const multisense::MultiSenseConfig &config, const multisense::MultiSenseInfo& info)
+{
+    //
+    // Sensor resolution
+
+    std::string valid_resolutions = "valid_resolutions [width, height, max disparity value]\n";
+    for (const auto& device_mode : info.operating_modes)
+    {
+        valid_resolutions += "[";
+        valid_resolutions += std::to_string(device_mode.width) + ", ";
+        valid_resolutions += std::to_string(device_mode.height) + ", ";
+        valid_resolutions += std::to_string(disparity_pixels(device_mode.disparities)) + "]\n";
+    }
+
+    rcl_interfaces::msg::ParameterDescriptor sensor_resolution_desc;
+    sensor_resolution_desc.set__name("sensor_resolution")
+                          .set__type(rcl_interfaces::msg::ParameterType::PARAMETER_INTEGER_ARRAY)
+                          .set__description(valid_resolutions);
+
+    declare_parameter("sensor_resolution",
+                      std::vector<int64_t>{config.width, config.height, disparity_pixels(config.disparities)},
+                      sensor_resolution_desc);
+
+    //
+    // Imu rates and ranges
+
+    if (info.imu)
+    {
+        if (info.imu->accelerometer && config.imu_config && config.imu_config->accelerometer)
+        {
+            create_imu_parameters(info.imu->accelerometer.value(),
+                                  config.imu_config->accelerometer.value(),
+                                  "imu.accelerometer");
+        }
+        if (info.imu->gyroscope && config.imu_config && config.imu_config->gyroscope)
+        {
+            create_imu_parameters(info.imu->gyroscope.value(),
+                                  config.imu_config->gyroscope.value(),
+                                  "imu.gyroscope");
+        }
+        if (info.imu->magnetometer && config.imu_config && config.imu_config->magnetometer)
+        {
+            create_imu_parameters(info.imu->magnetometer.value(),
+                                  config.imu_config->magnetometer.value(),
+                                  "imu.magnetometer");
+        }
+    }
+
+    //
+    // If our main image pair does not support color, make sure we explicitly disable white balance. If
+    // we could undeclare the parameter
```

Comment:
is this an unfinished comment?

looks like parameters can be undeclared by setting to rclcpp::PARAMETER_NOT_SET?